src/meta/src/hummock/manager/mod.rs

```diff
@@ -983,58 +1039,83 @@ where
             }
         }
 
-        let task_status = compact_task.task_status();
         debug_assert!(
-            task_status != TaskStatus::Pending,
+            compact_task.task_status() != TaskStatus::Pending,
             "report pending compaction task"
         );
-        if let TaskStatus::Success = task_status {
+        {
             // The compaction task is finished.
             let mut versioning_guard = write_lock!(self, versioning).await;
             let versioning = versioning_guard.deref_mut();
             let current_version = &mut versioning.current_version;
-            let mut hummock_version_deltas =
-                BTreeMapTransaction::new(&mut versioning.hummock_version_deltas);
-            let version_delta = gen_version_delta(
-                &mut hummock_version_deltas,
-                current_version,
-                compact_task,
-                CompactStatus::is_trivial_move_task(compact_task),
-                deterministic_mode,
-            );
+            let is_success = if let TaskStatus::Success = compact_task.task_status() {
+                let is_expired = match current_version
+                    .get_levels()
+                    .get(&compact_task.compaction_group_id)
+                {
+                    Some(levels) => Self::is_compact_task_expired(compact_task, levels),
+                    None => true,
+                };
+                if is_expired {
+                    compact_task.set_task_status(TaskStatus::InvalidGroupCanceled);
+                    false
+                } else {
+                    true
+                }
+            } else {
+                false
+            };
+            if is_success {
+                let mut hummock_version_deltas =
+                    BTreeMapTransaction::new(&mut versioning.hummock_version_deltas);
+                let mut branched_ssts = BTreeMapTransaction::new(&mut compaction.branched_ssts);
```

Comment:
It means that we must maintains all sstableinfos in `branched_ssts`  even if most of them would only be reference by one group? I think this would cause a lot write-keys on etcd because we must insert a key for every sstable file.

Reply:
In fact, `HummockVersion` has record the reference relationship if we collect this info from every group. So we only need to record `branched_ssts` in memory rather than store it in ETCD.

Reply:
> It means that we must maintains all sstableinfos in `branched_ssts` even if most of them would only be reference by one group? I think this would cause a lot write-keys on etcd because we must insert a key for every sstable file.

?
I thought that `branched_ssts` will only maintain SST that has been divided.
in `gen_version_delta`, we do not insert anything into `branched_ssts`

Reply:
> In fact, `HummockVersion` has record the reference relationship if we collect this info from every group. So we only need to record `branched_ssts` in memory rather than store it in ETCD.

yes. Although `branched_ssts` do not contain all SSTs, I still recommend recording only in memory. I will do some refactor. thanks

Reply:
> In fact, `HummockVersion` has record the reference relationship if we collect this info from every group. So we only need to record `branched_ssts` in memory rather than store it in ETCD.

I just modified code. Now it's only in memory